File: api/recognize.js
const crypto = require('crypto');

const MAX_IMAGE_BYTES = 3 * 1024 * 1024;
const API_URLS = {
  standard: 'https://server.simpletex.cn/api/latex_ocr',
  turbo: 'https://server.simpletex.cn/api/latex_ocr_turbo',
};
const IMAGE_TYPES = {
  'image/png': { extension: 'png', type: 'image/png' },
  'image/jpeg': { extension: 'jpg', type: 'image/jpeg' },
  'image/webp': { extension: 'webp', type: 'image/webp' },
};

function signedHeaders(appId, appSecret) {
  const headers = {
    'app-id': appId,
    'random-str': crypto.randomBytes(12).toString('base64url'),
    timestamp: String(Math.floor(Date.now() / 1000)),
  };
  const signatureSource = Object.keys(headers).sort().map((key) => `${key}=${headers[key]}`).join('&');
  headers.sign = crypto.createHash('md5').update(`${signatureSource}&secret=${appSecret}`).digest('hex');
  return headers;
}

function sendJson(response, status, body) {
  response.setHeader('Cache-Control', 'no-store');
  response.status(status).json(body);
}

module.exports = async function handler(request, response) {
  if (request.method !== 'POST') return sendJson(response, 405, { error: 'Method not allowed.' });
  const { appId, appSecret, image, imageType, model = 'standard' } = request.body || {};
  if (typeof appId !== 'string' || typeof appSecret !== 'string' || !appId.trim() || !appSecret.trim()) return sendJson(response, 400, { error: '请填写 SimpleTex APP ID 与 APP Secret。' });
  if (!API_URLS[model] || !IMAGE_TYPES[imageType] || typeof image !== 'string') return sendJson(response, 400, { error: '请求格式无效。' });

  let imageBuffer;
  try { imageBuffer = Buffer.from(image, 'base64'); } catch { return sendJson(response, 400, { error: '图片数据无效。' }); }
  if (!imageBuffer.length || imageBuffer.length > MAX_IMAGE_BYTES) return sendJson(response, 400, { error: '图片不能超过 3 MB。' });

  const imageInfo = IMAGE_TYPES[imageType];
  const form = new FormData();
  form.append('file', new Blob([imageBuffer], { type: imageInfo.type }), `formula.${imageInfo.extension}`);
  try {
    const upstream = await fetch(API_URLS[model], { method: 'POST', headers: signedHeaders(appId.trim(), appSecret.trim()), body: form, signal: AbortSignal.timeout(25_000) });
    const payload = await upstream.json();
    if (!upstream.ok || !payload.status) return sendJson(response, upstream.status || 502, { error: simpletexError(upstream.status, payload) });
    const latex = payload.res?.latex;
    if (!latex) return sendJson(response, 502, { error: 'SimpleTex 未返回 LaTeX 结果。' });
    return sendJson(response, 200, { latex, confidence: payload.res.conf, requestId: payload.request_id });
  } catch (error) {
    const message = error.name === 'TimeoutError' ? '连接 SimpleTex 超时，请稍后重新识别。' : '无法连接 SimpleTex，请检查网络后重试。';
    return sendJson(response, 502, { error: message });
  }
};

function simpletexError(status, payload) {
  const messages = {
    req_unauthorized: '鉴权失败，请检查 APP ID 与 APP Secret。',
    resource_no_valid: '没有可用识别额度，请检查 SimpleTex 账户余额或资源包。',
    image_missing: '未收到图片，请重新上传或粘贴。',
    image_oversize: '图片过大，请压缩后重试。',
    exceed_max_qps: '请求过于频繁，请稍后重试。',
    exceed_max_ccy: '当前请求过多，请稍后重试。',
    sever_closed: 'SimpleTex 服务维护中，请稍后重试。',
  };
  return messages[payload.errType || payload.error_type] || `SimpleTex 识别失败（HTTP ${status}）。`;
}

File: api/recognize.py
import base64
import hashlib
import json
import secrets
import string
import time
from http.server import BaseHTTPRequestHandler

import requests


# JSON/base64 expands an image by roughly one third. Keep the full request below
# Vercel Functions' 4.5 MB payload limit.
MAX_IMAGE_BYTES = 3 * 1024 * 1024
IMAGE_TYPES = {
    "image/png": ("formula.png", "image/png"),
    "image/jpeg": ("formula.jpg", "image/jpeg"),
    "image/webp": ("formula.webp", "image/webp"),
}
SIMPLETEX_API_URLS = {
    "standard": "https://server.simpletex.cn/api/latex_ocr",
    "turbo": "https://server.simpletex.cn/api/latex_ocr_turbo",
}
SIGN_CHARS = string.ascii_letters + string.digits


def simpletex_headers(data: dict, app_id: str, app_secret: str) -> dict[str, str]:
    headers = {
        "app-id": app_id,
        "random-str": "".join(secrets.choice(SIGN_CHARS) for _ in range(16)),
        "timestamp": str(int(time.time())),
    }
    values = {**data, **headers}
    sign_source = "&".join(f"{key}={values[key]}" for key in sorted(values))
    headers["sign"] = hashlib.md5(
        f"{sign_source}&secret={app_secret}".encode()
    ).hexdigest()
    return headers


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != "/api/recognize":
            self.respond(404, {"error": "Not found."})
            return

        try:
            content_length = int(self.headers.get("Content-Length", "0"))
            if not 0 < content_length <= MAX_IMAGE_BYTES * 2:
                raise ValueError("图片过大或请求无效。")
            body = json.loads(self.rfile.read(content_length))
            app_id = body["appId"].strip()
            app_secret = body["appSecret"].strip()
            model = body.get("model", "standard")
            image_type = body["imageType"]
            image = base64.b64decode(body["image"], validate=True)
            if not app_id or not app_secret:
                raise ValueError("请填写 SimpleTex APP ID 与 APP Secret。")
            if model not in SIMPLETEX_API_URLS:
                raise ValueError("不支持的识别模型。")
            if image_type not in IMAGE_TYPES:
                raise ValueError("不支持的图片格式。")
            if not 0 < len(image) <= MAX_IMAGE_BYTES:
                raise ValueError("图片不能超过 3 MB。")
        except (KeyError, TypeError, ValueError, json.JSONDecodeError) as exc:
            self.respond(400, {"error": str(exc) or "请求格式无效。"})
            return

        try:
            response = requests.post(
                SIMPLETEX_API_URLS[model],
                data={},
                files={"file": (IMAGE_TYPES[image_type][0], image, IMAGE_TYPES[image_type][1])},
                headers=simpletex_headers({}, app_id, app_secret),
                timeout=25,
            )
            payload = response.json()
            if not isinstance(payload, dict):
                raise ValueError
        except requests.Timeout:
            self.respond(504, {"error": "连接 SimpleTex 超时，请稍后重新识别。"})
            return
        except (requests.RequestException, ValueError):
            self.respond(502, {"error": "SimpleTex 返回了无效响应，请稍后重试。"})
            return

        if not response.ok or not payload.get("status"):
            self.respond(response.status_code or 502, {"error": simpletex_error(response.status_code, payload)})
            return

        result = payload.get("res", {})
        latex = result.get("latex")
        if not latex:
            self.respond(502, {"error": "SimpleTex 未返回 LaTeX 结果。"})
            return
        self.respond(200, {
            "latex": latex,
            "confidence": result.get("conf"),
            "requestId": payload.get("request_id"),
        })

    def respond(self, status: int, payload: dict):
        encoded = json.dumps(payload, ensure_ascii=False).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json; charset=utf-8")
        self.send_header("Content-Length", str(len(encoded)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(encoded)


def simpletex_error(status: int, payload: dict) -> str:
    messages = {
        "req_unauthorized": "鉴权失败，请检查 APP ID 与 APP Secret。",
        "resource_no_valid": "没有可用识别额度，请检查 SimpleTex 账户余额或资源包。",
        "image_missing": "未收到图片，请重新上传或粘贴。",
        "image_oversize": "图片过大，请压缩后重试。",
        "exceed_max_qps": "请求过于频繁，请稍后重试。",
        "exceed_max_ccy": "当前请求过多，请稍后重试。",
        "sever_closed": "SimpleTex 服务维护中，请稍后重试。",
        "server_inference_error": "SimpleTex 推理失败，请重新识别或切换模型。",
        "image_proc_error": "图片处理失败，请换一张清晰的图片重试。",
    }
    return messages.get(payload.get("errType") or payload.get("error_type"), {
        401: "鉴权失败，请检查 APP ID 与 APP Secret。",
        402: "没有可用识别额度，请检查 SimpleTex 账户余额或资源包。",
        413: "图片缺失或过大，请重新上传或压缩后重试。",
        429: "请求过于频繁，请稍后重试。",
        503: "SimpleTex 服务维护中，请稍后重试。",
    }.get(status, f"SimpleTex 识别失败（HTTP {status}）。"))
